src/freezing/freezing_apply_changes.py

```diff
@@ -1,156 +1,289 @@
+            )
+
+    # apply additions
+    if additions_exist:
+        additions_df = read_csv(additions_filepath)
+        if additions_df.empty:
+            freezing_logger.warning(
+            f"Additions file {additions_filepath} is empty, skipping..."
+        )
+        else:
+            additions_df["instance"] = additions_df["instance"].astype("Int64")
+            main_df = apply_additions(
+                main_df, 
+                additions_df, 
+                run_id, 
+                freezing_logger
+            )
+
+    return main_df
+
+
+def validate_any_refinst_in_frozen(
+        frozen_df: pd.DataFrame,
+        df2: pd.DataFrame,
+    ) -> bool:
+    """Validate that any of the ref/inst combinations from df2 are in the frozen df.
+
+    Args:
+        frozen_df (pd.DataFrame): The frozen csv df
+        df2 (pd.DataFrame): A second dataframe.
+
+    Returns:
+        bool: Whether any ref/inst combs from df2 are in frozen_df.
+    """
+    frozen_copy = frozen_df.copy()
+    df2_copy = df2.copy()
+    frozen_copy["refinst"] = (
+        frozen_copy["reference"].astype(str) + frozen_copy["instance"].astype(str)
+    )
+    df2_copy["refinst"] = (
+        df2_copy["reference"].astype(str) + df2_copy["instance"].astype(str)
+    )
+    result = any([x in frozen_copy["refinst"] for x in df2_copy["refinst"]])
+    return result
+
+
+def validate_all_refinst_in_frozen(
+        frozen_df: pd.DataFrame,
+        df2: pd.DataFrame,
+    ) -> bool:
+    """Validate that all ref/inst combinations in a list are in a df.
+
+    Args:
+        frozen_df (pd.DataFrame): The frozen csv df.
+        df2 (pd.DataFrame): The ammendments/additions df.
+
+    Returns:
+        bool: Whether all ref/inst are in the dataframe. 
+    """
+    frozen_copy = frozen_df.copy()
+    frozen_copy["refinst"] = (
+        frozen_copy["reference"].astype(str) + frozen_copy["instance"].astype(str)
+    )
+    result = values_in_column(
+        frozen_copy,
+        "refinst",
+        df2["reference"].astype(str) + df2["instance"].astype(str)
+    )
+    return result
+
+
+def validate_amendments_df(
+        frozen_df: pd.DataFrame,
+        amendments_df: pd.DataFrame,
+        freezing_logger: logging.Logger,
+    ) -> bool:
+    """Validate the amendments df.
+
+    Args:
+        frozen_df (pd.DataFrame): The frozen csv df.
+        amendments_df (pd.DataFrame): The amendments df.
+        freezing_logger (logging.Logger): The logger to log to.
+
+    Returns:
+        bool: Whether or not the amendments df is valid.
+    """
+    # check that all ref/inst combs are in staged frozen data
+    freezing_logger.info(
+        "Checking if all ref/inst in the amendments df are present in the frozen"
+        " data..."
+    )
+    present = validate_all_refinst_in_frozen(frozen_df, amendments_df)
+    if not present: 
+        freezing_logger.info(
+            "Not all reference/instance combinations found within the amendments"
```

Comment:
Amendments isn't working as expected - it's being skipped because it's failing validation, but I've been through and all ref/instances are present